addons/product_expiry/models/production_lot.py

```diff
@@ -29,51 +30,35 @@ def _compute_product_expiry_alert(self):
                 lot.product_expiry_alert = False
 
-    def _get_dates(self, product_id=None):
-        """Returns dates based on number of days configured in current lot's product."""
-        mapped_fields = {
-            'expiration_date': 'expiration_time',
-            'use_date': 'use_time',
-            'removal_date': 'removal_time',
-            'alert_date': 'alert_time'
-        }
-        res = dict.fromkeys(mapped_fields, False)
-        product = self.env['product.product'].browse(product_id) or self.product_id
-        if product:
-            for field in mapped_fields:
-                duration = getattr(product, mapped_fields[field])
-                if duration:
-                    date = datetime.datetime.now() + datetime.timedelta(days=duration)
-                    res[field] = fields.Datetime.to_string(date)
-        return res
-
-    # Assign dates according to products data
-    @api.model_create_multi
-    def create(self, vals_list):
-        for vals in vals_list:
-            dates = self._get_dates(vals.get('product_id') or self.env.context.get('default_product_id'))
-            for d in dates:
-                if not vals.get(d):
-                    vals[d] = dates[d]
-        return super().create(vals_list)
-
-    @api.onchange('expiration_date')
-    def _onchange_expiration_date(self):
-        if not self._origin or not (self.expiration_date and self._origin.expiration_date):
-            return
-        time_delta = self.expiration_date - self._origin.expiration_date
-        # As we compare expiration_date with _origin.expiration_date, we need to
-        # use `_get_date_values` with _origin to keep a stability in the values.
-        # Otherwise it will recompute from the updated values if the user calls
-        # this onchange multiple times without save between each onchange.
-        vals = self._origin._get_date_values(time_delta)
-        self.update(vals)
-
-    @api.onchange('product_id')
-    def _onchange_product(self):
-        dates_dict = self._get_dates()
-        for field, value in dates_dict.items():
-            setattr(self, field, value)
+    @api.depends('product_id')
+    def _compute_expiration_date(self):
+        for lot in self:
+            if not lot.product_id.use_expiration_date:
+                lot.expiration_date = False
+            elif not lot.expiration_date:
+                duration = lot.product_id.product_tmpl_id.expiration_time
+                lot.expiration_date = datetime.datetime.now() + datetime.timedelta(days=duration)
+
+    @api.depends('product_id', 'expiration_date')
+    def _compute_dates(self):
+        for lot in self:
+            if not lot.product_id.use_expiration_date:
+                lot.use_date = False
+                lot.removal_date = False
+                lot.alert_date = False
+            elif lot.expiration_date:
+                # when create
+                if not lot.use_date and not lot.removal_date and not lot.alert_date:
```

Comment:
I think we should recompute `use_date`, `removal_date` and `alert_date` also when we change the product:
```suggestion
                if lot.product_id != lot._origin.product_id or\
                   (not lot.use_date and not lot.removal_date and not lot.alert_date):
```
If you create a lot for a product using expiration dates, than change the lot's product for another product also using expiration date, applying the time delta on the existing dates doesn't make sense since we don't know if the two products are configured in the same way.
@thomasdeleval Do you agree ?